From 0aa58a9c225a1408bc6f35ded46e3e450b59ada2 Mon Sep 17 00:00:00 2001
From: Leymooo <vjatseslav.maspanov@gmail.com>
Date: Fri, 2 Apr 2021 01:22:06 +0300
Subject: [PATCH] Rebranding


diff --git a/build.gradle b/build.gradle
index a1395e47..4c64ab93 100644
--- a/build.gradle
+++ b/build.gradle
@@ -17,7 +17,7 @@ allprojects {
     apply plugin: "com.github.spotbugs"
 
     group 'com.velocitypowered'
-    version '1.1.5-SNAPSHOT'
+    version 'botfilter-1.0-SNAPSHOT'
 
     ext {
         // dependency versions
diff --git a/proxy/build.gradle b/proxy/build.gradle
index 33fa5bb4..ef444b6e 100644
--- a/proxy/build.gradle
+++ b/proxy/build.gradle
@@ -29,7 +29,7 @@ jar {
         }
 
         attributes 'Main-Class': 'com.velocitypowered.proxy.Velocity'
-        attributes 'Implementation-Title': "Velocity"
+        attributes 'Implementation-Title': "Velocity-BotFilter"
         attributes 'Implementation-Version': version
         attributes 'Implementation-Vendor': "Velocity Contributors"
         attributes 'Multi-Release': 'true'
diff --git a/settings.gradle b/settings.gradle
index d0f3098b..9c6d1e92 100644
--- a/settings.gradle
+++ b/settings.gradle
@@ -1,4 +1,4 @@
-rootProject.name = 'velocity'
+rootProject.name = 'velocity-botfilter'
 include(
         'api',
         'proxy',
-- 
2.29.2.windows.3

